File: androidApp/src/main/java/io/redlink/more/app/android/observations/Polar/Polar360UserProfile.kt
package io.redlink.more.app.android.observations.Polar

import android.content.Context
import com.polar.sdk.api.model.PolarFirstTimeUseConfig
import io.redlink.more.app.android.MoreApplication
import java.util.Calendar

data class Polar360UserProfile(
    val gender: Gender,
    val age: Int,
    val heightCm: Int,
    val weightKg: Int
) {
    enum class Gender(val displayName: String) {
        FEMALE("Female"),
        MALE("Male");

        val polarGender: PolarFirstTimeUseConfig.Gender
            get() = when (this) {
                FEMALE -> PolarFirstTimeUseConfig.Gender.FEMALE
                MALE -> PolarFirstTimeUseConfig.Gender.MALE
            }
    }

    /** Approximate birth date derived from age (Jan 1 of the birth year). */
    val birthDate: java.util.Date
        get() {
            val cal = Calendar.getInstance()
            cal.set(Calendar.YEAR, cal.get(Calendar.YEAR) - age)
            cal.set(Calendar.MONTH, Calendar.JANUARY)
            cal.set(Calendar.DAY_OF_MONTH, 1)
            return cal.time
        }

    companion object {
        private const val PREFS_NAME = "polar360_profile"
        private const val KEY_GENDER = "gender"
        private const val KEY_AGE = "age"
        private const val KEY_HEIGHT = "heightCm"
        private const val KEY_WEIGHT = "weightKg"

        private fun prefs(context: Context) =
            context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

        fun load(): Polar360UserProfile? {
            val ctx = MoreApplication.appContext ?: return null
            val prefs = prefs(ctx)
            val genderRaw = prefs.getString(KEY_GENDER, null) ?: return null
            val gender = runCatching { Gender.valueOf(genderRaw) }.getOrNull() ?: return null
            if (!prefs.contains(KEY_AGE) || !prefs.contains(KEY_HEIGHT) || !prefs.contains(KEY_WEIGHT)) return null
            return Polar360UserProfile(
                gender = gender,
                age = prefs.getInt(KEY_AGE, 0),
                heightCm = prefs.getInt(KEY_HEIGHT, 0),
                weightKg = prefs.getInt(KEY_WEIGHT, 0)
            )
        }

        fun save(profile: Polar360UserProfile) {
            val ctx = MoreApplication.appContext ?: return
            prefs(ctx).edit()
                .putString(KEY_GENDER, profile.gender.name)
                .putInt(KEY_AGE, profile.age)
                .putInt(KEY_HEIGHT, profile.heightCm)
                .putInt(KEY_WEIGHT, profile.weightKg)
                .apply()
        }
    }
}

File: iosApp/iosApp/Observations/Polar360UserProfile.swift
//
//  Polar360UserProfile.swift
//  iosApp
//

import Foundation
import PolarBleSdk

struct Polar360UserProfile {

    enum Gender: String, CaseIterable {
        case female, male

        var displayName: String {
            switch self {
            case .female: return "Female"
            case .male: return "Male"
            }
        }

        var polarGender: PolarFirstTimeUseConfig.Gender {
            switch self {
            case .female: return .female
            case .male: return .male
            }
        }
    }

    var gender: Gender
    var age: Int
    var heightCm: Int
    var weightKg: Int

    /// Approximate birth date derived from age (set to Jan 1 of the birth year).
    var birthDate: Date {
        let year = Calendar.current.component(.year, from: Date()) - age
        return Calendar.current.date(from: DateComponents(year: year, month: 1, day: 1)) ?? Date()
    }

    // MARK: - Persistence

    private static let genderKey = "polar360.profile.gender"
    private static let ageKey    = "polar360.profile.age"
    private static let heightKey = "polar360.profile.heightCm"
    private static let weightKey = "polar360.profile.weightKg"

    static func load() -> Polar360UserProfile? {
        guard
            let defaults = AppDelegate.appGroupUserDefaults,
            let genderRaw = defaults.string(forKey: genderKey),
            let gender = Gender(rawValue: genderRaw),
            defaults.object(forKey: ageKey) != nil,
            defaults.object(forKey: heightKey) != nil,
            defaults.object(forKey: weightKey) != nil
        else { return nil }

        return Polar360UserProfile(
            gender: gender,
            age: defaults.integer(forKey: ageKey),
            heightCm: defaults.integer(forKey: heightKey),
            weightKg: defaults.integer(forKey: weightKey)
        )
    }

    func save() {
        guard let defaults = AppDelegate.appGroupUserDefaults else { return }
        defaults.set(gender.rawValue, forKey: Self.genderKey)
        defaults.set(age, forKey: Self.ageKey)
        defaults.set(heightCm, forKey: Self.heightKey)
        defaults.set(weightKg, forKey: Self.weightKey)
    }
}
